cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/MetricValues.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ */
+
+package com.linkedin.cruisecontrol.monitor.sampling.aggregator;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.StringJoiner;
+
+
+/**
+ * A class hosting the values of a particular metric.
+ */
+public class MetricValues implements Serializable {
+
+  private final float[] _values;
+  private volatile float _max;
+  private volatile double _sumForAvg;
+
+  public MetricValues(int numWindows) {
+    _values = new float[numWindows];
+    _sumForAvg = 0;
+    _max = Float.MIN_VALUE;
+  }
+
+  public void set(int index, double value) {
+    if (_max == _values[index] && value < _max) {
+      _max = -1.0f;
+    }
+    _sumForAvg += value - _values[index];
+    _values[index] = (float) value;
+  }
+
+  public double get(int index) {
+    return _values[index];
+  }
+
+  public void clear() {
+    Arrays.fill(_values, 0);
+    _sumForAvg = 0;
+    _max = Float.MIN_VALUE;
+  }
+
+  public int length() {
+    return _values.length;
+  }
+
+  public void add(double[] values) {
+    _max = Float.MIN_VALUE;
+    for (int i = 0; i < _values.length; i++) {
+      double toAdd = values[i];
+      _values[i] += toAdd;
```

Comment:
Even adding a float to a float may overflow. When have floating point calculations, we usually do not worry about overflow because the range is so big, for a float, the range is at 10^38.